### PR TITLE
Fix Issue #256

### PR DESCRIPTION
**Pull request by br111an**
OxyPlot.Series.HeatMapSeries was calculating a nonsense value when
double.NaN was present. This is issue #256. Fix interpolation to avoid calculating nonsense.
Labels use integer coordinates so a non-interpolating method was
introduced to get values for integer coordinates.

### CODE REVIEW

File: Source/OxyPlot/Series/HeatMapSeries.cs [hunk 1]
```diff
@@ -405,34 +407,54 @@ private static double GetValue(double[,] data, double i, double j)
...
-                    return closestData;
-                }
+            double ifraction = i - i0;
+            double jfraction = j - j0;
+            var v0 = (data[i0, j0] * (1 - ifraction)) + (data[i1, j0] * ifraction);
+            var v1 = (data[i0, j1] * (1 - ifraction)) + (data[i1, j1] * ifraction);
 
-                return v0 * (1 - jx);
-            }
+            ////if (double.IsNaN(v0))
```
Comment (objorke) [inline Source/OxyPlot/Series/HeatMapSeries.cs]: delete commented code

File: Source/OxyPlot/Series/HeatMapSeries.cs [hunk 2]
```diff
@@ -405,34 +407,54 @@ private static double GetValue(double[,] data, double i, double j)
...
+
+            ////    return v0 * (1 - jx);
+            ////}
+
+            return (v0 * (1 - jfraction)) + (v1 * jfraction);
+        }
+
+        /// <summary>
+        /// Labels are not interpolated so we have a method that accepts integer coordinates.
+        /// This addresses Issue 256.
```
Comment (objorke) [inline Source/OxyPlot/Series/HeatMapSeries.cs]: remove this line - no need to refer to github issues in the code

**Comment by objorke** [comment]
+ Update the change log / contributors files
+ Is there an example for this issue? If not, add one!
+ Use a hash in PR description to get the issure reference right: #256 

:+1:

**Comment by br111an** [comment]
How do I hash the issue reference?  (Issue 256)

Brian

**Comment by br111an** [comment]
Example tests added.
Pull request title modified to put the hash mark ahead of the issue number.
Change log updated.
Contributor list updated.

**Comment by br111an** [comment]
I believe I addressed your concerns appropriately. Please advise if there is more that I need to do for Issue #256.

**Comment by objorke** [comment]
Please include #560 and #561 in this pull request. Can you please also rebase and squash to clean up the commit history?

**Comment by objorke** [comment]
Please remove the FixIssue256 project, or move it to an appropriate place (Source\Examples\WPF\WpfExamples). I think the examples added in the ExampleLibrary should be sufficient. Only add platform specific examples where functionality cannot be demonstrated in the example library.
I prefer to keep the number of applications at a minimum to reduce maintenance work and build times... (the examples in the WpfExamples project was refactored from 34 applications to 1...)

File: Source/OxyPlot/Series/HeatMapSeries.cs [hunk 3]
```diff
@@ -405,34 +406,32 @@ private static double GetValue(double[,] data, double i, double j)
...
 
-            return (v0 * (1 - jx)) + (v1 * jx);
+        /// <summary>
+        /// Labels are not interpolated so we have a method that accepts integer coordinates.
+        /// </summary>
+        /// <param name="data">The data.</param>
+        /// <param name="i">The first index.</param>
+        /// <param name="j">The second index.</param>
+        /// <returns>A double that might be double.NaN.</returns>
+        private static double GetValueForLabel(double[,] data, int i, int j)
```
Comment (objorke) [inline Source/OxyPlot/Series/HeatMapSeries.cs]: why add this static method? I don't understand :-)

**Comment by objorke** [comment]
Thanks for updating! It looks 99% good now! Please update the changelog, I suggest "Show NaN values correctly in HeatMapSeries (#256)" under the fixed section. Then squash all the commits into one.
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html
https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

**Comment by br111an** [comment]
Issue #256 is actually about side-effects of NaN so I suggest "Show values correctly when proximate to NaN in HeatMapSeries (#256).".

**Comment by br111an** [comment]
Please have another look. GetValue has been improved where coordinates are whole numbers. Previously, both (i, j) had to be whole numbers to receive special treatment. Now the whole number check is for each dimension independently. Note: a double equal to an integer will cast to that integer because the mantissa of a double is large enough. Reference: http://stackoverflow.com/questions/9326027/

**Comment by objorke** [comment]
Good improvement! 
Consider to move the code in the static GetValue method to a general linear inteprolaiton "utility" class. Then add some unit tests to make sure it works correctly.
Please update the change log and squash the commits - then I can merge!

**Comment by br111an** [comment]
I did the rebase and squash.

**Comment by objorke** [comment]
Did you push the rebased and squashed commit? Your `develop` branch is still based on code from December 2014. The change log file was added in January 2015.

**Comment by br111an** [comment]
I don't know how to do that. Please pull it so I don't have to push it.

**Comment by br111an** [comment]
I deleted br111an/oxyplot. I will try to start over again.

**Comment by br111an** [comment]
I forked and cloned again. This time the change log file was present. Edited the change log in the "fixed" section. Added myself to the contributors file. Fixed the bug of issue #256. Added test code featuring heat maps of flat data that better illustrate the bug. Commited. Synched to github.

The single commit obviates the need to re-base and squash so I did not follow these instructions at your link.

http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Your link does not instruct to use a push command. I hope I'm finished with issue #256 and that you do not require anything else for this issue.

**Comment by objorke** [comment]
your branch https://github.com/br111an/oxyplot/commits/develop looks good! please create a PR, and I will merge

**Pull request closed without merging**